pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -62,14 +62,37 @@ type ResolvedPipelineRunTask struct {
 // state of the PipelineRun.
 type PipelineRunState []*ResolvedPipelineRunTask
 
+func (state PipelineRunState) IsDone() (isDone bool) {
+	isDone = true
+	for _, t := range state {
+		if t.TaskRun == nil {
+			return false
+		}
+		status := t.TaskRun.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
+		retriesDone := len(t.TaskRun.Status.RetriesStatus)
+		retries := t.PipelineTask.Retries
+		println(status, retries, retriesDone)
+		isDone = isDone && (status.IsTrue() || status.IsFalse() && retriesDone > retries)
+		if !isDone {
+			return
+		}
+	}
+	return
+}
+
 // GetNextTasks will return the next ResolvedPipelineRunTasks to execute, which are the ones in the
 // list of candidateTasks which aren't yet indicated in state to be running.
-func (state PipelineRunState) GetNextTasks(candidateTasks map[string]v1alpha1.PipelineTask) []*ResolvedPipelineRunTask {
+func (state *PipelineRunState) GetNextTasks(candidateTasks map[string]v1alpha1.PipelineTask) []*ResolvedPipelineRunTask {
 	tasks := []*ResolvedPipelineRunTask{}
-	for _, t := range state {
+	for _, t := range *state {
 		if _, ok := candidateTasks[t.PipelineTask.Name]; ok && t.TaskRun == nil {
 			tasks = append(tasks, t)
 		}
+		if _, ok := candidateTasks[t.PipelineTask.Name]; ok && t.TaskRun != nil {
+			if len(t.TaskRun.Status.RetriesStatus) < t.PipelineTask.Retries {
```

Comment:
im not certain but i think you should check here if the candidate task has failed - that's the only case we'd want to retry it in (we don't want to retry if it is still executing, or it succeed)

^^ and i think we need some more test cases for the above